### PR TITLE
add compiler docs

### PR DESCRIPTION
**Pull request by lidaamber**
Add compiler docs and examples.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
 // ...
 connection.event("interfaceName", "methodName", args);
 ```
+
+### JSTP compiler
+
+JSTP compiler is a nice feature for better handling of JSTP packets. You can declare interfaces comfortable for concrete handling purposes and use them instead of `ManualHandler`.
```
Comment (lundibundi) [inline README.md]: I think it's better to put it smth like
JSTP compiler is a nice feature to ease handling of JSTP packets. You can declare interfaces that correspond to specific API or simple call to avoid writing all boilerplate code to get the arguments out of the packet. JSTP compiler will parse those at compile time and generate implementations.

File: README.md [hunk 2]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+
+#### Installation
+
+[Check for the latest version](https://bintray.com/metarhia/maven/jstp-compiler)
+
+Gradle:
+```
+dependencies {
+  compile 'com.metarhia.jstp:jstp-compiler:0.1.12'
```
Comment (lundibundi) [inline README.md]: I think it's better to use new [gradle syntax](https://docs.gradle.org/current/userguide/artifact_dependencies_tutorial.html) for compile and other methods or add both

File: README.md [hunk 3]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+```
+You can declare your own methods that will be called in `onInvoke()` method of basic ManualHandler.
+
+##### Named
+
+Gets the field of the received packet by specified name. Can be multi-level.
+
+```java
+@JSTPHandler
+public interface ExampleHandler {
```
Comment (lundibundi) [inline README.md]: Let's rename it to OkErrorHandler?

File: README.md [hunk 4]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+  void onError(JSArray args);
+
+  // gets value by key "neededValue" in object got by "ok"
+  @Named(value = {"ok", "neededValue"})
+  void onNeededValueGot(String value);
+  // ...
+}
+```
+
+##### NotNull
```
Comment (lundibundi) [inline README.md]: Should be before the previous paragraph as it is used there.

File: README.md [hunk 5]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+
+Checks if the needed value is not null before calling the handler method.
```
Comment (lundibundi) [inline README.md]: We should elaborate that the method will not be called if value is null.

File: README.md [hunk 6]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+  // ...
+  @NotNull
+  void onExampleValue(JSArray args);
+  // ...
+}
+```
+
+##### Indexed
+
+Can be used for getting concrete argument from the received arguments. Can be multi-level
```
Comment (lundibundi) [inline README.md]: for getting -> to get. Or maybe it's better to change the phrase as it has too much 'argument' in it =).

File: README.md [hunk 7]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+@JSTPHandler
+public interface ExampleHandler {
+  // ...
+}
+```
+You can declare your own methods that will be called in `onInvoke()` method of basic ManualHandler.
+
+##### Named
+
+Gets the field of the received packet by specified name. Can be multi-level.
```
Comment (lundibundi) [inline README.md]: Maybe better smth like
Can be multi-level. -> Obviously allows getting elements from nested objects, the keys will be getted in the order specified

File: README.md (hunk 6 above)
Comment (lundibundi) [inline README.md]: 'Can be multi-level' same as before, it's better to elaborate it.

File: README.md [hunk 8]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+
+  // gets packet[1][2]
+   void onValueBySecondIndex(@Indexed(value = {1,2}) JSArray args);
+  // ...
+}
+```
+
+##### Custom-named
+
+It is a sort of combination of `Named` and `Indexed` annotations. You can get needed value by index or by key. Can be multi-level
```
Comment (lundibundi) [inline README.md]: 'Can be multi-level' same as before.

Comment (lundibundi) [inline README.md]: I think we should explain the syntax here (about [1], {1}, "").

File: README.md [hunk 9]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+#### JSTP receiver
+You can process received values not only via single handler, but by several ones. They can be added or removed to certain `JSTP receiver`, generated similarly to `JSTPHandler`. To generate receiver, you need to do the following:
+
+```java
+@JSTPReceiver
+public interface ExampleReceiver {
+  // ...
+}
+```
+You can declare required methods similarly to `JSTPHandler`. After compilation class `JSTPExampleReceiver` will be generated and you will be able to use it in packets processing.
```
Comment (lundibundi) [inline README.md]: I'd rather write that syntax is the same and not similar.

File: README.md [hunk 10]
```diff
@@ -296,3 +296,147 @@ JSArray args = new JSArray();
...
+  void onNeededValue(JSValue args);
+
+  // gets packet["ok"][1][2]
+  @CustomNamed(value = {"ok", "[1]", "{2}"})
+  void onNeededValue(JSValue args);
+  // ...
+}
+```
+
+After compilation class `JSTPExampleHandler` will be generated and you will be able to use it in packets processing.
```
Comment (lundibundi) [inline README.md]: packets processing -> packet processing ?

File: README.md (hunk 1 above)
Comment (lidaamber) [inline README.md]: Yes, that'll be ok as well

File: README.md (hunk 2 above)
Comment (lidaamber) [inline README.md]: Do you mean using the full definition instead of shortcut?

File: README.md (hunk 4 above)
Comment (lidaamber) [inline README.md]: Ok

File: README.md (hunk 8 above)
Comment (lidaamber) [inline README.md]: Ok

File: README.md (hunk 9 above)
Comment (lidaamber) [inline README.md]: Ok

File: README.md (hunk 2 above)
Comment (lundibundi) [inline README.md]: I meant to add such definition
`compile group: 'com.metarhia.jstp', name: 'jstp-compiler', version: '0.1.12`'
here as well as in the other section.

File: README.md [hunk 11]
```diff
@@ -47,7 +47,8 @@ JSNativeSerializer.stringify(arr); // returns "[1,2,3]"
 ```
 If it doesn't know how to serialize input it'll be serialized as `undefined`
 
-2) To simple js mirrored hierarchy in java with `JSParser` (hierarchy has `JSValue` as superclass)
+2) To simple js mirrored hierarchy in java with `JSParser` (hierarchy has
```
Comment (lundibundi) [inline README.md]: This should've been in a different pr or commit as they are not related to this one. But let's leave as it is now.

File: README.md [hunk 12]
```diff
@@ -100,8 +101,10 @@ try {
 }
 ```
 
-To convert values from js java hierarchy to js use `.toString()` method or `JSTP.stringify`.
-They can be parsed in js with a simple eval statement or [js parser](https://github.com/metarhia/JSTP)
+To convert values from js java hierarchy to js use `.toString()` method or
```
Comment (lundibundi) [inline README.md]: Same as before.

File: README.md [hunk 13]
```diff
@@ -333,48 +339,54 @@ public interface ExampleHandler {
   // ...
 }
 ```
-You can declare your own methods that will be called in `onInvoke()` method of basic ManualHandler.
+You can declare your own methods that will be called in `onInvoke()` method of
```
Comment (lundibundi) [inline README.md]: I think that we should avoid this statement as it reveals unneeded implementation details. And instead write the usages of JSTPHandler at the beginning of this section.

File: README.md [hunk 14]
```diff
@@ -333,48 +339,54 @@ public interface ExampleHandler {
...
+basic ManualHandler.
 
-##### Named
+##### NotNull
 
-Gets the field of the received packet by specified name. Can be multi-level.
+Checks if the needed value is not null before calling the handler method
```
Comment (lundibundi) [inline README.md]: Let write it as 'Annotates that the method should only be called if all of its arguments are not null (in case of method-wide annotation) or only specific parameters are not null (in case of argument-wide annotations)'

File: README.md [hunk 15]
```diff
@@ -333,48 +339,54 @@ public interface ExampleHandler {
...
+  void onExampleValue(JSArray args);
   // ...
 }
 ```
 
-##### NotNull
+##### Named
 
-Checks if the needed value is not null before calling the handler method.
+Gets the field of the received packet by specified name. It also allows getting
```
Comment (lundibundi) [inline README.md]: 'It also allows getting elements from nested obj, the value will be retrieved in the order of keys specified.'
Thanks to @LeraSavchenko.

File: README.md [hunk 16]
```diff
@@ -333,48 +339,54 @@ public interface ExampleHandler {
...
+  @Named(value = {"ok", "neededValue"})
+  void onNeededValueGot(String value);
   // ...
 }
 ```
 
 ##### Indexed
 
-Can be used for getting concrete argument from the received arguments. Can be multi-level
+Can be used to get concrete argument from the received ones. It also allows
```
Comment (lundibundi) [inline README.md]: maybe 'specific value from jstp message'?

Comment (lundibundi) [inline README.md]: 'It also ...' adapt from the previous comment.

File: README.md [hunk 17]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
 
 ##### Custom-named
 
-It is a sort of combination of `Named` and `Indexed` annotations. You can get needed value by index or by key. Can be multi-level
+It is a sort of combination of `Named` and `Indexed` annotations. You can get
+needed value by index or by key. It also allows getting elements from nested
+objects and arrays, the value will be got in the order the keys and indexes
```
Comment (lundibundi) [inline README.md]: 'the value will be got...' same as before.

File: README.md [hunk 18]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
...
+are specified. To get a value by key, you should just declare the required key
+like in `@Named` annotation. To get an array by index, you can declare it as
+`[index]`, in such a way you can get values from this array like `{inner index}`.
```
Comment (lundibundi) [inline README.md]: We should wrap [index] and {inner index} with quotes as those are strings.

File: README.md [hunk 19]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
 
 ##### Custom-named
 
-It is a sort of combination of `Named` and `Indexed` annotations. You can get needed value by index or by key. Can be multi-level
+It is a sort of combination of `Named` and `Indexed` annotations. You can get
+needed value by index or by key. It also allows getting elements from nested
+objects and arrays, the value will be got in the order the keys and indexes
+are specified. To get a value by key, you should just declare the required key
+like in `@Named` annotation. To get an array by index, you can declare it as
```
Comment (lundibundi) [inline README.md]: an array -> value from array

File: README.md (hunk 18 above)
Comment (lundibundi) [inline README.md]: 'this array'? With such annotation they'll be retrieved by the nth(inner index) key of the object.

File: README.md [hunk 20]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
...
 public interface ExampleHandler {
   // ...
 
   @CustomNamed("ok")
-  void onNeededValue(JSValue args);
+  void onNeededNamedValue(JSValue args);
 
   @CustomNamed("{1}")
-  void onNeededValue(JSValue args);
+  void onNeededIndexValue(JSValue args);
```
Comment (lundibundi) [inline README.md]: Key by index?

File: README.md [hunk 21]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
...
 
 ```java
 connection.call("interfaceName", "methodName", args, new JSTPExampleHandler() {
     // ...
 });
 ```
 #### JSTP receiver
-You can process received values not only via single handler, but by several ones. They can be added or removed to certain `JSTP receiver`, generated similarly to `JSTPHandler`. To generate receiver, you need to do the following:
+You can process received values not only via single handler, but by several
+ones. They can be added or removed to certain `JSTP receiver`, generated
```
Comment (lundibundi) [inline README.md]: to certain `JSTP receiver` -> from `JSTP receiver` via \<proper method names\>.
And I think 'generated..' should be separate sentence as it is not related to this one.

File: README.md [hunk 22]
```diff
@@ -390,43 +402,55 @@ public interface ExampleHandler {
...
+similarly to `JSTPHandler`. To generate receiver, you need to do the following:
 
 ```java
 @JSTPReceiver
 public interface ExampleReceiver {
   // ...
 }
 ```
-You can declare required methods similarly to `JSTPHandler`. After compilation class `JSTPExampleReceiver` will be generated and you will be able to use it in packets processing.
+The syntax of declaring methods is the same to `JSTPHandler`. After compilation
```
Comment (lundibundi) [inline README.md]: to -> as in ?

File: README.md [hunk 23]
```diff
@@ -360,7 +363,7 @@ public interface ExampleHandler {
 ##### Named
 
 Gets the field of the received packet by specified name. It also allows getting
-elements from nested objects, the value will be got in the order the keys are
+elements from nested objects, the value will be retrieved in the order the keys
```
Comment (lundibundi) [inline README.md]: Grammar is not fixed, see previous comment.

File: README.md [hunk 24]
```diff
@@ -377,16 +380,16 @@ public interface OkErrorHandler {
...
 }
 ```
 
 ##### Indexed
 
-Can be used to get concrete argument from the received ones. It also allows
-getting elements from nested arrays, the value will be got in the order the
-indexes are specified.
+Can be used to get the specific value from JSTP message. It also allows
+getting elements from nested arrays, the value will be retrieved in the order the
```
Comment (lundibundi) [inline README.md]: Same as before.

File: README.md [hunk 25]
```diff
@@ -404,10 +407,11 @@ public interface ExampleHandler {
 
 It is a sort of combination of `Named` and `Indexed` annotations. You can get
 needed value by index or by key. It also allows getting elements from nested
-objects and arrays, the value will be got in the order the keys and indexes
-are specified. To get a value by key, you should just declare the required key
-like in `@Named` annotation. To get an array by index, you can declare it as
-`[index]`, in such a way you can get values from this array like `{inner index}`.
+objects and arrays, the value will be retrieved in the order the keys and indexes
```
Comment (lundibundi) [inline README.md]: Same grammar mistake.

**Comment by lundibundi** [comment]
Landed in d619d2c75babdf463562e4a11308498b958fa136.
